### PR TITLE
Reshuffle internal targets and add HPX::hpx_no_wrap_main target

### PR DESCRIPTION
**Pull request by msimberg**
"Fixes"  #4537 and makes it a bit clearer which targets are internal and which are not.

Adds a layer of indirection to the `HPX::hpx` target so that we can have a version which does not link to `hpx_wrap`. All public targets are now in the `HPX::` namespace and all others are either in `HPXInternal::` or no namespace (the ones that are created in `HPXConfig.cmake`.

~The `hpx` library target is temporarily renamed `hpx_some_name` in search of a better name, because `hpx` is taken by the new interface target.~ The old `hpx` target is called `hpx_core` now. The actual library name is still `libhpx.so` (or the equivalent on Windows).

The `hpx_init` and `hpx_wrap` merge from #4462 has also been undone here.

I do have a nagging feeling that this is backwards and that we should make `hpx_wrap` opt-in, rather than opt-out, but that's something we shouldn't change now and can consider if and when the time comes for HPX 2.

### CODE REVIEW

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -103,15 +103,15 @@ endif()
 
 ################################################################################
 # libhpx
-add_library(hpx ${hpx_library_link_mode_core}
+add_library(hpx_some_name ${hpx_library_link_mode_core}
```
Comment (aurianer) [inline src/CMakeLists.txt:106]: `hpx_intermediate` ? `hpx_inter` ?

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -218,24 +218,37 @@ if(MSVC AND HPX_WITH_SWAP_CONTEXT_EMULATION)
...
-target_link_libraries(hpx_interface INTERFACE $<${_is_executable}:$<TARGET_NAME_IF_EXISTS:HPXInternal::hpx_init>>)
-target_compile_definitions(hpx_interface INTERFACE "$<${_is_executable}:HPX_APPLICATION_NAME_DEFAULT=$<TARGET_PROPERTY:NAME>>")
-target_compile_definitions(hpx_interface INTERFACE "$<${_is_executable}:HPX_PREFIX_DEFAULT=\"${HPX_PREFIX}\">")
-target_compile_definitions(hpx_interface INTERFACE "$<${_is_executable}:HPX_APPLICATION_EXPORTS>")
-target_compile_definitions(hpx_interface INTERFACE "$<${_is_library}:HPX_LIBRARY_EXPORTS>")
+target_link_libraries(hpx_interface INTERFACE hpx_interface_no_wrap_main)
+target_link_libraries(hpx_interface INTERFACE $<${_is_executable}:$<TARGET_NAME_IF_EXISTS:hpx_wrap>>)
+target_link_libraries(hpx_interface INTERFACE $<${_is_executable}:$<TARGET_NAME_IF_EXISTS:HPXInternal::hpx_wrap>>)
+
+target_link_libraries(hpx_no_wrap_main INTERFACE hpx_interface_no_wrap_main)
```
Comment (aurianer) [inline src/CMakeLists.txt:251]: Add `hpx_no_wrap_main` to `hpx_targets` otherwise it doesn't get exported

File: src/CMakeLists.txt (hunk 1 above)
Comment (msimberg) [reply src/CMakeLists.txt:106]: I may just call it `hpx_core` for now. The name will be hidden from users so we can change it easily in the future.

File: src/CMakeLists.txt (hunk 2 above)
Comment (msimberg) [reply src/CMakeLists.txt:251]: Good catch!

**Comment by hkaiser** [comment]
@msimberg Please note that in order to solve #4580 I have added special linker flags to executables by associating it with `hpx_init` (https://github.com/STEllAR-GROUP/hpx/blob/master/init/CMakeLists.txt#L103-L110). So far I don't see this in this PR. Should it go somewhere else now?

**Comment by msimberg** [comment]
> @msimberg Please note that in order to solve #4580 I have added special linker flags to executables by associating it with `hpx_init` (https://github.com/STEllAR-GROUP/hpx/blob/master/init/CMakeLists.txt#L103-L110). So far I don't see this in this PR. Should it go somewhere else now?

I've left that unchanged: https://github.com/STEllAR-GROUP/hpx/pull/4564/files#diff-534e69f2776a2ab1d04f5bc130b2a4caL103. I think it can stay there unless I come up with a good reason to move it.

**Pull request merged**